### ISSUE TITLE
Enhancement: allow custom ESTree parser service

### ISSUE THREAD

**Issue by JLHwung** [typescript-eslint/typescript-eslint#10878]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

The goal of this issue is to enable developers to try out new ECMAScript features in a TypeScript codebase without complete loss of the linter support.

We are aligning Babel's TypeScript AST to typescript-eslint in Babel 8 (https://github.com/babel/babel/issues/16679). However, there is no current solution to integrate Babel parser with this project.

This feature is not limited to babel-eslint, instead it can serve for broader use case: People can make their own parser for a new proposal (either JS or TS) and experiment it on a codebase.

babel-eslint should handle experimental JS features such that this project can work. The current linter rules will not support new experimental AST nodes, e.g. records, tuples, structs, etc, unless they have reached stage 4. However, the linter itself should tolerate unknown AST nodes from a third-party parser: it should not throw on new AST nodes. If we want to support linter within the new nodes: the visitor keys should accept augmentation.

### Additional Info

For example, we can introduce a `customParser` option:

```js
// eslint.config.js
import { parseForTypescriptESLint } from "@babel/eslint-parser"
import typescriptEslint from "typescript-eslint";
export default [
  {
    languageOptions: {
      parser: typescriptEslint.parser,
      parserOptions: {
        customParser: parseForTypescriptESLint
      },
    },
  }
]
```

Here `parseForTypescriptESLint` should return a typescript-estree compatible AST. This option probably does not work with the `projectService` because we can't intercept the `tsc` parsing. Of course this is a tentative design and feel free to revise it.

**Comment by JoshuaKGoldberg** [comment]
> This option probably does not work with the `projectService` because we can't intercept the `tsc` parsing

Yeah 😬 typed linting requires a TypeScript-project-produced AST. This holds true for both `project` and `projectService. If TypeScript can't parse a file then we don't have any way of providing type information for it.

On the other hand, typed linting is the only area that requires TypeScript AST nodes. You can already use a custom parser with non-typed typescript-eslint rules. In essence, this _should_ be "fine" as long as the AST produced by the parser matches the TSESTree shape:

```ts
// eslint.config.js
import { parseForTypescriptESLint } from "@babel/eslint-parser";
import eslint from '@eslint/js';
import tseslint from 'typescript-eslint';

export default [
  eslint.configs.recommended,
  tseslint.configs.recommended,
  {
    languageOptions: {
      parser: parseForTypescriptESLint
    },
  }
]
```

Example past issue showing a rule being decoupled from parser services: https://github.com/typescript-eslint/typescript-eslint/issues/9860

I think this is a combination of:

* Already working: for the untyped case, since it should already work 
* Blocked/wontfix: for the typed case, since we're blocked on TS parsing

Did I understand right?

**Comment by JLHwung** [comment]
Thank you for the prompt reply and the detailed example. It is very helpful. From what I understand, a custom ESLint parser should return `{ ast, scopeManager, visitorKeys }`:

For `ast`, if current `typescript-eslint/parser` [customized options](https://typescript-eslint.io/packages/parser#configuration) are all `tsc`-related, then it looks good to me to directly replace the parser to be the one provided from `babel-eslint`, otherwise `babel-eslint` will have to copy every supported non-typing features from the tseslint parser, which is not very efficient and maintainable.

For `scopeManager` and `visitorKeys`, it seems that we can directly inherit `@typescript-eslint/scope-manager` and `@typescript-eslint/visitor-keys`, right?

**Comment by JoshuaKGoldberg** [comment]
I haven't looked closely at `babel-eslint`, but IIRC it does have some AST differences. You'll probably need to verify that each rule runs correctly with it - we certainly don't recommend users use it. https://github.com/typescript-eslint/typescript-eslint/issues/8150

Otherwise 👍 looks good to me.

**Comment by bradzacher** [comment]
The major issue right now is the differences in AST yeah.
There are many places that babel does not align with us. Here's a list of all of our test fixtures where babel differs from us:
https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/ast-spec/tests/fixtures-with-differences-ast.shot
